src/main/java/ServerConfig.java

```diff
@@ -1,11 +1,23 @@
 class ServerConfig {
   final int port;
+  final int maxConnections;
+  final int readTimeout;
+
+  static final int DEFAULT_PORT = 80;
+  static final int DEFAULT_MAX_CONNECTIONS = 50;
+  static final int DEFAULT_READ_TIMEOUT = 10000;
 
   ServerConfig() {
-    this.port = 80;
+    this(DEFAULT_PORT, DEFAULT_MAX_CONNECTIONS, DEFAULT_READ_TIMEOUT);
```

Comment:
Nice use of constructor delegation here. I do wonder if all these variants are necessary. I see that the `ServerConfig` is injected into the `TcpServer` constructor (great!), and that the "real" usage of this config, in `TcpServerAcceptanceTest`, specifies all 3 arguments, including one default that's in the middle! 

This suggests to me that you might want one of:
- a lot more constructor overloads, with these arguments ordered in every permutation - 😿
- a `Map` - 😐- maybe premature but not bad - if/when you introduce a config file or loading these via system properties, this is probably where you'd want to go. Then the benefits of this class would basically just be defaults and type safety.
- a single constructor that requires the client to specify all of these settings - 🙂- this would be my personal preference for now, because it's not yet clear to me that the defaults are useful.

Reply:
My preferred approach is to use the [builder pattern](https://dzone.com/articles/design-patterns-the-builder-pattern). However. it's incredibly cumbersome to implement by hand, which is why I would look to use a library the one I had before (which generated the `Builder` and `Immutable*` counterparts to your value type automatically).

Having said that, I'd prefer living with the current setup until the delegating constructors become obsolete or unused. Here are my thoughts on the suggestions you made:

> a lot more constructor overloads
I don't think you need all the permutations. That said, Java's lack of keyword arguments makes this sort of thing even more tedious.

> a Map - 😐- maybe premature but not bad 
I'm not a fan of `Map` in the programatic API because it eschews static checking of required arguments.

> a single constructor
...to rule them all! Also OK with me, but I don't think it's worth removing the "convenience" constructors just yet. I can see their demise in a future PR though... 😈